### PR TITLE
[critical] fix Issue 16530 - -O -cov interaction leads to wrong codegen

### PR DESCRIPTION
**Pull request by WalterBright**
This also seems to have fixed a tab issue in test42.d.

Anyhow, this is a critical bug fix.

### CODE REVIEW

File: src/backend/cgcod.c [hunk 1]
```diff
@@ -2292,20 +2292,23 @@ STATIC code * comsub(elem *e,regm_t *pretregs)
...
+        if (regcon.cse.value[i] == e)   // if contents are right
+                emask |= mask[i];       // turn on bit for reg
+    }
+    emask &= regcon.cse.mval;                     // make sure all bits are valid
 
-  /* First construct a mask, emask, of all the registers that   */
-  /* have the right contents.                                   */
+    if (emask & XMMREGS && *pretregs == mPSW)
+        ;
+    else if (tyfloating(e->Ety) && config.inline8087)
```
Comment (andralex) [inline src/backend/cgcod.c]:
```d
if (((emask & XMMREGS) == 0 || *pretregs != mPSW) && tyfloating(e->Ety) && config.inline8087)
    return comsub87(e,pretregs);
```

I guess the config check could be put first, too, as it's cheapest.

**Comment by UplinkCoder** [comment]
@WalterBright please split the formatting changes in another PR.
It's hard to make out the changes.

**Review by andralex — COMMENTED** [review]
Cool beans thx. One nit?

**Comment by andralex** [comment]
yo quiero git stash

File: src/backend/cgcod.c (hunk 1 above)
Comment (WalterBright) [inline src/backend/cgcod.c]: Except config.inline8087 is almost always on.

**Comment by WalterBright** [comment]
> please split the formatting changes in another PR.

Guess I deserved that one. Done.

**Review by UplinkCoder — APPROVED** [review]

File: src/backend/cgcod.c (hunk 1 above)
Comment (andralex) [inline src/backend/cgcod.c]: cool, then the suggestion remains as is

**Comment by andralex** [comment]
eh I'll just pull this in spite of my awesome suggestion not getting in

**Comment by andralex** [comment]
Auto-merge toggled on

**Pull request merged**